src/luascript.cpp

```diff
@@ -13777,66 +13812,927 @@ int LuaScriptInterface::luaPartySetSharedExperience(lua_State* L)
+			return 1;
+		}
+		spell = g_spells->getRuneSpellByName(arg);
+		if (spell) {
+			pushUserdata<RuneSpell>(L, dynamic_cast<RuneSpell*>(spell));
+			setMetatable(L, -1, "Spell");
+			return 1;
+		}
+	} else {
+		lua_pushnil(L);
+	}
+	return 1;
+}
+
+int LuaScriptInterface::luaSpellOnCastSpell(lua_State* L)
+{
+	// spell:onCastSpell(callback)
+	Spell* spell = dynamic_cast<InstantSpell*>(getUserdata<InstantSpell>(L, 1));
+	if (spell) {
+		if (spell->spellType != SPELL_UNDEFINED) {
+			InstantSpell* instant = getUserdata<InstantSpell>(L, 1);
+			if (!instant->loadCallback()) {
+				pushBoolean(L, false);
+				return 1;
+			}
+			instant->scripted = true;
+			pushBoolean(L, true);
+		} else {
+			RuneSpell* rune = getUserdata<RuneSpell>(L, 1);
+			if (!rune->loadCallback()) {
+				pushBoolean(L, false);
+				return 1;
+			}
+			rune->scripted = true;
+			pushBoolean(L, true);
+		}
+	} else {
+		lua_pushnil(L);
+	}
+	return 1;
+}
+
+int LuaScriptInterface::luaSpellRegister(lua_State* L)
+{
+	// spell:register()
+	Spell* spell = dynamic_cast<InstantSpell*>(getUserdata<InstantSpell>(L, 1));
+	if (spell) {
+		if (spell->spellType == SPELL_UNDEFINED) {
+			spell = dynamic_cast<RuneSpell*>(getUserdata<RuneSpell>(L, 1));
+		}
+
+		if (spell->spellType == SPELL_INSTANT) {
+			InstantSpell* instant = getUserdata<InstantSpell>(L, 1);
+			if (!instant->isScripted()) {
+				pushBoolean(L, false);
+				return 1;
+			}
+			pushBoolean(L, g_spells->registerInstantLuaEvent(instant));
+		} else if (spell->spellType == SPELL_RUNE) {
+			RuneSpell* rune = getUserdata<RuneSpell>(L, 1);
+			if (rune->getMagicLevel() != 0 || rune->getLevel() != 0) {
+				//Change information in the ItemType to get accurate description
+				ItemType& iType = Item::items.getItemType(rune->getRuneItemId());
+				iType.name = rune->getName();
+				iType.runeMagLevel = rune->getMagicLevel();
+				iType.runeLevel = rune->getLevel();
+				iType.charges = rune->getCharges();
+			}
+			if (!rune->isScripted()) {
+				pushBoolean(L, false);
+				return 1;
+			}
+			pushBoolean(L, g_spells->registerRuneLuaEvent(rune));
+		}
 	} else {
-		std::string stringArgument = getString(L, 2);
-		spell = g_spells->getInstantSpellByName(stringArgument);
-		if (!spell) {
-			spell = g_spells->getInstantSpell(stringArgument);
+		lua_pushnil(L);
+	}
+	return 1;
+}
+
+int LuaScriptInterface::luaSpellName(lua_State* L)
+{
+	// spell:name(name)
+	Spell* spell = dynamic_cast<InstantSpell*>(getUserdata<InstantSpell>(L, 1));
+	if (spell) {
+		if (spell->spellType == SPELL_UNDEFINED) {
+			spell = dynamic_cast<RuneSpell*>(getUserdata<RuneSpell>(L, 1));
+		}
+
+		if (lua_gettop(L) == 1) {
+			pushString(L, spell->getName());
+		} else {
+			spell->setName(getString(L, 2));
+			pushBoolean(L, true);
 		}
+	} else {
+		lua_pushnil(L);
 	}
+	return 1;
+}
 
+int LuaScriptInterface::luaSpellId(lua_State* L)
+{
+	// spell:id(id)
+	Spell* spell = dynamic_cast<InstantSpell*>(getUserdata<InstantSpell>(L, 1));
```

Comment:
This `dynamic_cast` seems redundant.

Reply:
well I tried to initialize as `InstantSpell* without `dynamic_cast` and if it was not the right `spellType` I tried to `dynamic_cast` sideways onto `RuneSpell*` that however didn't work as intended and gave sometimes UB.
I didn't feel like making more pointers, would just bulk up the entire code actually, yet again if you have a better solution for this then feel free to let me know

Reply:
The `dynamic_cast` doesn't save the day here if the pointer we retrieve is not an `InstantSpell`, the pointer we get is an "invalid pointer" (http://eel.is/c++draft/expr.static.cast#2).
`dynamic_cast<T*>(p)` requires that `p` is either `nullptr` or a valid pointer.

Here's an [example](https://godbolt.org/z/bfHaT0). Both clang and gcc optimize out the `if` check, because of UB caused by the `static_cast` in line 21.

Reply:
fixed in [9e888a9](https://github.com/otland/forgottenserver/pull/2584/commits/9e888a96bb55d97ef102cb18a8c91e78178c8ecf)